### ISSUE TITLE
[AC3] readQuery returns a whole new copy of results after a mutation updates a subset of results.

### ISSUE THREAD

**Issue by ashubham** [apollographql/apollo-client#6249]
**Intended outcome:**

If a `mutation` updates a subset of result, the siblings of the changed subset should still be the same reference of data as before.

**Actual outcome:**

When running `readQuery` after a partial update of the result set due to a mutation, the result set has all children as cloned copies even though they have not changed.

**How to reproduce the issue:**

My schema looks something like:
```
type Mutation {
  search(query: string): SearchCompletions!
}

type Query {
  getAnswer(session: Session!): AnswerSession!
}

type Session {
  id: ID!
  genNo: Int // This increments on every mutation.
}

type AnswerSession {
  session: Session!
  answers: [Answer!]
}

type SearchCompletions {
  session: Session!
  completions: [string!]
}

type Answer {
  // some fields here ...
}
```

Whenever I do a `search` mutation, it returns with `CompletionResult`, which updates the Session in the client cache with the new `genNo`.

Now, if do a `Query` with `getAnswer` query, the `answers` array is always a deep clone of the existing set even though it hasnt changed.

This is causing my react components to rerender.

**Versions**

System:
    OS: macOS 10.15.4
  Binaries:
    Node: 12.16.2 - /usr/local/bin/node
    npm: 6.14.4 - /usr/local/bin/npm
  Browsers:
    Chrome: 81.0.4044.129
    Firefox: 75.0
    Safari: 13.1
  Apollo Client: 3.0-beta.45

**Comment by ashubham** [comment]
I did more research on this, seems to be due to `dirtyChildren` check which is done in `QueryManager.ts`
https://github.com/apollographql/apollo-client/blob/master/src/cache/inmemory/readFromStore.ts#L98

Basically, we `executeSelectionSet` if any child is marked `dirty`. In the above case `Session` is a child of `AnswerSession` which is marked dirty, causing the whole `AnswerSession` to be re-executed leading to new objects for `answers` key.

**Comment by ashubham** [comment]
@benjamn Is this something we are tracking to fix in 3.0 ? Or is this something expected

**Comment by benjamn** [comment]
This is working as (currently) designed.

Although we track result caching dependencies at the field level, the "results" saved by the result caching system are entity objects, not individual fields, so the whole entity object gets reread when any of its fields change.

It's tempting to cache individual field values, because then we could just invalidate field values without invalidating the whole entity object, but in practice caching individual fields slows down reading and writing, and uses a lot more memory, since most fields are relatively cheap to compute (and thus not worth caching).

I'm happy to revisit these tradeoffs in the future (after AC3 is released), but that's the current reasoning.
